### PR TITLE
VP9 support

### PR DESCRIPTION
**Pull request by sadjad**

### CODE REVIEW

File: src/aiortc/codecs/vpx.py [hunk 1]
```diff
@@ -436,6 +436,275 @@ def __update_config(self) -> None:
...
+            if bitrate != self.__target_bitrate:
+                self.__target_bitrate = bitrate
+                self.__update_config_needed = True
+
+    def __update_config(self) -> None:
+        self.cfg.rc_target_bitrate = self.__target_bitrate // 1000
+        self.__update_config_needed = False
+
 
 def vp8_depayload(payload: bytes) -> bytes:
```
Comment (vibhaa) [inline src/aiortc/codecs/vpx.py:709]: Is the payload descriptor identical between vp8/vp9?

File: src/aiortc/codecs/vpx.py [hunk 2]
```diff
@@ -436,6 +436,275 @@ def __update_config(self) -> None:
...
+                        f"buf_optimal_sz {str(self.cfg.rc_buf_optimal_sz)}, buf_sz {str(self.cfg.rc_buf_sz)}, "\
+                        f"buf_initial_sz {str(self.cfg.rc_buf_initial_sz)}, ")
+            self.__update_config()
+            _vpx_assert(lib.vpx_codec_enc_init(self.codec, self.cx, self.cfg, 0))
+
+            lib.vpx_codec_control_(
+                self.codec, lib.VP9E_SET_NOISE_SENSITIVITY, ffi.cast("int", 1)
+            )
+
+            lib.vpx_codec_control_(
```
Comment (vibhaa) [inline src/aiortc/codecs/vpx.py:598]: I assume these variables are identical?

File: src/aiortc/codecs/vpx.py [hunk 3]
```diff
@@ -436,6 +436,275 @@ def __update_config(self) -> None:
...
+                        i_pos += i_stride
+                        o_pos += o_stride
+
+                frames.append(frame)
+
+        return frames
+
+class Vp9Encoder(Encoder):
+    def __init__(self) -> None:
+        self.cx = lib.vpx_codec_vp9_cx()
```
Comment (vibhaa) [inline src/aiortc/codecs/vpx.py:501]: Same q as above: was this function already exposed or did you have to do something for it?

File: src/aiortc/codecs/vpx.py [hunk 4]
```diff
@@ -436,6 +436,275 @@ def __update_config(self) -> None:
         self.cfg.rc_target_bitrate = self.__target_bitrate // 1000
         self.__update_config_needed = False
 
+class Vp9Decoder(Decoder):
+    def __init__(self) -> None:
+        self.codec = ffi.new("vpx_codec_ctx_t *")
+        _vpx_assert(
+            lib.vpx_codec_dec_init(self.codec, lib.vpx_codec_vp9_dx(), ffi.NULL, 0)
```
Comment (vibhaa) [inline src/aiortc/codecs/vpx.py:443]: Was this already exposed or did you have to do something to expose this API?

File: src/aiortc/codecs/vpx.py [hunk 5]
```diff
@@ -436,6 +436,275 @@ def __update_config(self) -> None:
...
+                        f"max_quantizer {str(self.cfg.rc_max_quantizer)}, "\
+                        f"kf_max_dist {str(self.cfg.kf_max_dist)}, "\
+                        f"DEFAULT_BITRATE {self.__target_bitrate}, MIN_BITRATE {self.vpx_min_bitrate}, "\
+                        f"MAX_BITRATE {self.vpx_max_bitrate}, "\
+                        f"buf_optimal_sz {str(self.cfg.rc_buf_optimal_sz)}, buf_sz {str(self.cfg.rc_buf_sz)}, "\
+                        f"buf_initial_sz {str(self.cfg.rc_buf_initial_sz)}, ")
+            self.__update_config()
+            _vpx_assert(lib.vpx_codec_enc_init(self.codec, self.cx, self.cfg, 0))
+
+            lib.vpx_codec_control_(
```
Comment (vibhaa) [inline src/aiortc/codecs/vpx.py:594]: How did you know that this option needed to be set?

**Review by vibhaa — APPROVED** [review]
Just left a few questions for my own understanding.

File: src/aiortc/codecs/vpx.py (hunk 2 above)
Comment (sadjad) [reply src/aiortc/codecs/vpx.py:598]: Yes, some of the `VP8E_` variables are shared between VP8 and VP9.

File: src/aiortc/codecs/vpx.py (hunk 3 above)
Comment (sadjad) [reply src/aiortc/codecs/vpx.py:501]: This one was already exposed here: https://github.com/geminovc/aiortc/blob/main/src/_cffi_src/build_vpx.py#L213

File: src/aiortc/codecs/vpx.py (hunk 4 above)
Comment (sadjad) [reply src/aiortc/codecs/vpx.py:443]: Same as the `cx` function.

File: src/aiortc/codecs/vpx.py (hunk 5 above)
Comment (sadjad) [reply src/aiortc/codecs/vpx.py:594]: For VP8, it had already set a similar option, although it's a bit different for VP9: https://github.com/webmproject/libvpx/blob/main/vpx/vp8cx.h#L438

File: src/aiortc/codecs/vpx.py (hunk 1 above)
Comment (sadjad) [reply src/aiortc/codecs/vpx.py:709]: I thought the payload descriptor would be the same, but now that I'm looking at the RFCs ([7741](https://www.rfc-editor.org/rfc/rfc7741.html#section-4.2) for VP8 and [this draft](https://datatracker.ietf.org/doc/html/draft-ietf-payload-vp9#section-4.2) for VP9), it looks like they are different. Let me look into this a bit more, it shouldn't be hard to make the adjustments.

Comment (sadjad) [reply src/aiortc/codecs/vpx.py:709]: Okay, this works if both sides (the sender and the receiver) use the same payload format. This means that this implementation is NOT interoperable with a standard WebRTC implementation that uses the official payload format for VP9. This should be completely fine for training purposes.

Comment (sadjad) [reply src/aiortc/codecs/vpx.py:709]: My test code:

```python3
import os
import sys
import numpy as np
import matplotlib.pyplot as plt
import aiortc
import av
import importlib

from aiortc.codecs import vpx
from aiortc.jitterbuffer import JitterFrame

container = av.open(file='video.mp4', format=None, mode='r')
stream = container.streams.video[0]

encoder = vpx.Vp9Encoder()
decoder = vpx.Vp9Decoder()

for av_frame in container.decode(stream):
    payloads, ts = encoder.encode(av_frame, quantizer=52, enable_gcc=False)
    payload = b"".join([vpx.vp8_depayload(p) for p in payloads])
    decoded_frame_av = decoder.decode(JitterFrame(data=payload, timestamp=ts))
    f = decoded_frame_av[0].to_rgb().to_ndarray()
    plt.imshow(f)
    print(len(payload))
```

Comment (vibhaa) [reply src/aiortc/codecs/vpx.py:709]: Yeah, this may not be an immediate issue, but something worth looking into for end-to-end evaluation.

Comment (vibhaa) [reply src/aiortc/codecs/vpx.py:709]: Thank you for clarifying this, and sending me the test code.

File: src/aiortc/codecs/vpx.py (hunk 3 above)
Comment (vibhaa) [reply src/aiortc/codecs/vpx.py:501]: That's convenient!

File: src/aiortc/codecs/vpx.py (hunk 5 above)
Comment (vibhaa) [reply src/aiortc/codecs/vpx.py:594]: Interesting, thanks for the pointer!

**Pull request merged**